### PR TITLE
SSO State Migration + Logger refactor

### PR DESCRIPTION
**Pull request by iambmelt**
**Summary of changes:**

**API Changes:**
-  `PublicClientApplication#getAccounts()` is now asynchronous, performs ADAL token cache update if `0 == getAccounts().size() && !hasMigrated`
- Introduces new class `TokenMigrationUtility`, general purpose migration utility, leveraging `IShareSingleSignOnState`
- `IShareSingleSignOnState` now returns a `boolean` result indicated success/failure of SSO set-state

**Under the hood:**
    * Adds new `SharedPreferences` file to track state migration: `com.microsoft.identity.client.migration_status`
    * Does **not** delete any ADAL state after migration

**Corner cases + expected behavior:**
 - If upgrading from `1.15.+`, migration is skipped as the `MsalOAuth2TokenCache` is already synced by `ADALOAuth2TokenCache`
- Before attempting migration, the cache is consulted. If it contains >0 records, migration is skipped and it is marked as complete (so that it will not happen in the future if records are removed)
- Version rollbacks are supported because ADAL cache state is not deleted by migration
    * Please note: Any Accounts or Credentials added to an application _after_ a rollback will not be synced once `hasMigrated` is set to true.

**Corollary PR:**
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/296

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java [hunk 1]
```diff
@@ -307,6 +309,37 @@ public static String getSdkVersion() {
         return BuildConfig.VERSION_NAME;
     }
 
+    public void restoreAdalCache(final Context context) {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java]: This piece of API is going to go away

File: msal/src/main/java/com/microsoft/identity/client/TokenMigrationCallback.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.client;
+
+interface TokenMigrationCallback {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/TokenMigrationCallback.java]: Javadoc this

File: msal/src/main/java/com/microsoft/identity/client/IMigrationAdapter.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import android.util.Pair;
+
+import com.microsoft.identity.common.BaseAccount;
+import com.microsoft.identity.common.internal.providers.oauth2.RefreshToken;
+
+import java.util.List;
+import java.util.Map;
+
+public interface IMigrationAdapter<T extends BaseAccount, U extends RefreshToken> {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/IMigrationAdapter.java]: Javadoc this class

File: msal/src/main/java/com/microsoft/identity/client/AdalMigrationAdapter.java [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AdalMigrationAdapter implements IMigrationAdapter<MicrosoftAccount, MicrosoftRefreshToken> {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/AdalMigrationAdapter.java]: Add comments to this class - code is complex

File: msal/src/main/java/com/microsoft/identity/client/AdalMigrationAdapter.java [hunk 5]
```diff
@@ -0,0 +1,531 @@
+//  Copyright (c) Microsoft Corporation.
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/AdalMigrationAdapter.java]: Move to `common`

File: msal/src/main/java/com/microsoft/identity/client/IMigrationAdapter.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+//  Copyright (c) Microsoft Corporation.
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/IMigrationAdapter.java]: Move to `common`

File: msal/src/main/java/com/microsoft/identity/client/TokenMigrationUtility.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import com.microsoft.identity.common.BaseAccount;
+import com.microsoft.identity.common.internal.cache.IShareSingleSignOnState;
+import com.microsoft.identity.common.internal.providers.oauth2.RefreshToken;
+
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class TokenMigrationUtility<T extends BaseAccount, U extends RefreshToken> {
```
Comment (iambmelt) [inline msal/src/main/java/com/microsoft/identity/client/TokenMigrationUtility.java]: Move to `common`

File: testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/MainActivity.java [hunk 8]
```diff
@@ -246,46 +236,65 @@ public void onAcquireTokenClicked(final AcquireTokenFragment.RequestOptions requ
...
-
-        for (final IAccount accountToRemove : accountsToRemove) {
-            if (TextUtils.isEmpty(username) || accountToRemove.getUsername().equals(username.trim().toLowerCase())) {
-                    mApplication.removeAccount(accountToRemove);
+    public void onRemoveUserClicked(final String username) {
+        mApplication.getAccounts(new PublicClientApplication.AccountsLoadedListener() {
+            @Override
+            public void onAccountsLoaded(List<IAccount> accountsToRemove) {
+                for (final IAccount accountToRemove : accountsToRemove) {
+                    if (TextUtils.isEmpty(username) || accountToRemove.getUsername().equals(username.trim().toLowerCase())) {
```
Comment (addev-ashish) [inline testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/MainActivity.java]: how about using equalsIgnoreCase and removing the toLower, that way we get independent of the casing of acccountToRemove

File: testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/MainActivity.java [hunk 9]
```diff
@@ -246,46 +236,65 @@ public void onAcquireTokenClicked(final AcquireTokenFragment.RequestOptions requ
...
 
-        final IAccount requestAccount = getAccount(requestOptions.getLoginHint());
+        //final IAccount requestAccount = getAccount();
+        mApplication.getAccounts(new PublicClientApplication.AccountsLoadedListener() {
+            @Override
+            public void onAccountsLoaded(final List<IAccount> accounts) {
+                IAccount requestAccount = null;
+
+                for (final IAccount account : accounts) {
+                    if (account.getUsername().equals(requestOptions.getLoginHint().trim().toLowerCase())) {
```
Comment (addev-ashish) [inline testapps/testapp/src/main/java/com/microsoft/identity/client/testapp/MainActivity.java]: same here

**Pull request merged**